Add new guide for configuring mtls between gateway & authorino/limitador

david-martin opened the pull request:
This will be a new document in the 'Installation' section of https://docs.kuadrant.io
This documents the manual steps that were the outcome of an investigation detailed in https://github.com/Kuadrant/kuadrant-operator/issues/1049

A separate [RFC](https://github.com/Kuadrant/architecture/pull/110) is proposed for the possibility of a more automated way to handle enabling mTLS. 
This document is seen as a stopgap, should users desire to enable mTLS in the meantime.

maleck13 reviewed (commented):
This looks good. I am going to try it before I approve. 👍

eguzki commented:
Very nice! :lock: :100: :closed_lock_with_key:

guicassolato commented on doc/install/mtls-configuration.md:
```diff
@@ -0,0 +1,150 @@
+The next step enables mTLS for traffic originating in the gateway (where the Wasm plugin executes), going to the Kuadrant components.
+This requires modifying the EnvoyFilters directly.
+Here is an example `transport_socket` configuration to add to both the `kuadrant-auth-api-gateway` and `kuadrant-ratelimiting-api-gateway` EnvoyFilter resources.
+This should be added below `spec.configPatches[].patch.value`.
```
I wonder if "below" is unequivocal enough. To be more precise, I think we want to say the value of `spec.configPatches[].patch.value` (which is a YAML object) to be extended with this snippet below. I.e. for `transport_socket` to be added as a sibling of `name`, `type`, ..., `load_assignment`, is that right?

david-martin replied:
Yes. I'll udpate this to be clearer

guicassolato commented on doc/install/mtls-configuration.md:
````diff
@@ -0,0 +1,150 @@
+EOF
+```
+
+If you prefer to only able mTLS for a specific component, you can modify just the EnvoyFilter and Deployment for that component.
````
```suggestion
If you prefer to only enable mTLS for a specific component, you can modify just the EnvoyFilter and Deployment for that component.
```

guicassolato commented on doc/install/mtls-configuration.md:
```diff
@@ -0,0 +1,150 @@
+### Envoy Filter
+
+The next step enables mTLS for traffic originating in the gateway (where the Wasm plugin executes), going to the Kuadrant components.
+This requires modifying the EnvoyFilters directly.
```
I'm actually surprised this works. In theory, any events related to the following CRs should reconcile the EnvoyFilter back to no-TLS form:
- Kuadrant
- GatewayClass
- Gateway
- HTTPRoute
- AuthPolicy
- EnvoyFilter

david-martin replied:
I'll add a caveat that it may get undone, and they should allow for that by having some job or gitops thing reconcile it back.
I'll also link to the RFC where they can engage on it potential being a feature of kuadrant.

david-martin commented:
@guicassolato 
I've updated the words to make it clearer what has to be updated in EnvoyFilters.
I've also added a note about the changes potentially being undone, and a link to the rfc for people to engage if they'd like this as a feature.

@maleck13 
I've added some words to the overview note about ensuring service mesh is enabled, and a reference to Istio CNI (including a link to the sail operator docs).
There is also mention of having at least 1 AuthPolicy or RateLimitPolicy in the pre-reqs section (so an EnvoyFilter exists)
This has now been tested and verified on a local kind cluster and Openshift.
I know you had issues with the istio-proxy container coming up locally when setting up things a little differently that I had. Is that something you want to revisit or are you happy to approve the doc as it is?

maleck13 commented on doc/install/mtls-configuration.md:
```diff
@@ -0,0 +1,148 @@
+
+## Overview
+
+This guide includes manual steps to enable mTLS between the gateway and Kuadrant components.
```
```suggestion
This guide includes manual steps to enable mTLS between an Istio provided gateway and the Kuadrant components.
```

maleck13 commented on doc/install/mtls-configuration.md:
```diff
@@ -0,0 +1,148 @@
+
+This guide includes manual steps to enable mTLS between the gateway and Kuadrant components.
+If you use an AuthPolicy or RateLimitPolicy, there will be communication between the gateway and the respective Kuadrant components at request time. This communication happens between the Wasm plugin in Envoy proxy, and Authorino or Limitador.
+At the time of writing there is [an RFC](https://github.com/Kuadrant/architecture/pull/110) discussing how to add this capability as a feature of the Kuadrant operator. If you are interested in having that feature or influencing how it is delivered, please engage on that pull request.
```
```suggestion
At the time of writing there is [an RFC](https://github.com/Kuadrant/architecture/pull/110) discussing how to add mTLS capabilities as a feature of the Kuadrant operator. If you are interested in having that feature or influencing how it is delivered, please engage on that pull request.
```
